### PR TITLE
feat(file source): Add log namespace and schema def support

### PR DESCRIPTION
**Pull request by spencergilbert**
- save
- feat(file source): Add log namespace and schema def support
- add pub use
- feat(file source): Update Option<String> to Option<OptionalValuePath>
- feat(file source): Update host_key to Option<OptionalValuePath>

### CODE REVIEW

File: src/sources/file.rs [hunk 1]
```diff
@@ -467,13 +530,13 @@ pub fn file_source(
...
-            .unwrap_or_else(|| log_schema().host_key().to_string()),
+        host_key: config.host_key.clone().map_or_else(
+            || parse_value_path(log_schema().host_key()).ok(),
+            |key| key.path,
+        ),
         hostname: crate::get_hostname().ok(),
-        file_key: config.file_key.clone(),
-        offset_key: config.offset_key.clone(),
+        file_key: config.file_key.clone().and_then(|f| f.path),
+        offset_key: config.offset_key.clone().and_then(|f| f.path),
```
Comment (neuronull) [inline src/sources/file.rs:539]: mega nit but for consistency, could use `k.path` or `key.path` for all these

File: src/sources/file.rs [hunk 2]
```diff
@@ -659,35 +723,70 @@ fn wrap_with_line_agg(
...
+    log_namespace: LogNamespace,
+) -> LogEvent {
     emit!(FileEventsReceived {
         count: 1,
         file,
         byte_size: line.len(),
     });
 
-    let mut event = LogEvent::from_bytes_legacy(&line);
+    let deserializer = BytesDeserializer::new();
```
Comment (neuronull) [inline src/sources/file.rs:745]: Might have missed this but can you create the deserializer once (elsewhere), and pass it into this function?

**Review by neuronull — COMMENTED** [review]
Thanks for walking me through this, and for adding unit tests!

File: src/sources/file.rs (hunk 2 above)
Comment (spencergilbert) [reply src/sources/file.rs:745]: Probably? I followed the same pattern that I had already merged here: https://github.com/vectordotdev/vector/blob/master/src/sources/aws_s3/sqs.rs#L519

Comment (fuchsnj) [reply src/sources/file.rs:745]: `BytesDeserializer` is very cheap to create, so I don't think this one matters too much

Comment (neuronull) [reply src/sources/file.rs:745]: cool, yeah just noticed it on this PR.

File: src/sources/file.rs (hunk 1 above)
Comment (spencergilbert) [reply src/sources/file.rs]: https://github.com/vectordotdev/vector/pull/15174/commits/93702138ccbbf8b028752b0779d9161d74673575

**Review by neuronull — APPROVED** [review]

**Pull request merged**